### PR TITLE
Prevent finding rid in non-service entry spans

### PR DESCRIPTION
**Pull request by marcotc**
## Description

We are introducing proper `DD_TRACE_HEADER_TAGS` support in `dd-trace-rb` (https://github.com/DataDog/dd-trace-rb/pull/2935) and it trips a few appsec system-tests.

The issue happens because we tag non-service entry spans with `DD_TRACE_HEADER_TAGS`, but those tags are used to find the request id (`rid`) by system-tests to perform assertions.

It is likely very uncommon to set HTTP serve `DD_TRACE_HEADER_TAGS` tags in non-service entry spans, but it's the current state of the Ruby tracer. We should likely change this, and stop tagging Sinatra spans with `DD_TRACE_HEADER_TAGS`, but that's a breaking change for our clients and won't happen in the short term.
Disabling test until this change happens in the Ruby tracer is also not a great approach, because that would mean disabling a large set of appsec tests that we want to keep.

## Workflow

1. ⚠️⚠️ Create your PR as draft
2. Follow the style guidelines of this project (See [how to easily lint the code](https://github.com/DataDog/system-tests/blob/main/docs/edit/lint.md))
3. Work on you PR until the CI passes (if something not related to your task is failing, you can ignore it)
4. Mark it as ready for review

Once your PR is reviewed, you can merge it! :heart:

## Reviewer checklist

* [ ] Check what scenarios are modified. If needed, add the relevant label (`run-parametric-scenario`, `run-profiling-scenario`...). If this PR modifies any system-tests internal, then add the `run-all-scenarios` label ([more info](https://github.com/DataDog/system-tests/blob/main/docs/CI/labels.md)). 
* [ ] CI is green
   * [ ] If not, failing jobs are not related to this change (and you are 100% sure about this statement)
* if any of `build-some-image` label is present
  1. is the image labl have been updated ? 
  2. just before merging, locally build and push the image to hub.docker.com

### CODE REVIEW

**Comment by marcotc** [comment]
CI failures seem to mirror the master branch: https://github.com/DataDog/system-tests/actions/runs/5624028552

**Review by GustavoCaso — APPROVED** [review]
Love this 🎉 

Since it warranty that we are fetching the service level entry span when checking for `_dd.top_level` 👍 

Question: Should we apply the same check for the other conditions?

**Comment by marcotc** [comment]
> Question: Should we apply the same check for the other conditions?

I'm not confident it should, because the other conditions are documented as language-specific, and the for the `# last hope` check, we are likely looking to find anything useful at that point so having a wider-matching condition is probably ok.

File: utils/interfaces/_core.py [hunk 1]
```diff
@@ -170,14 +170,15 @@ def get_rid_from_span(span):
...
+    metrics = span.get("metrics", {})
 
     user_agent = None
 
     if span.get("type") == "rpc":
         user_agent = meta.get("grpc.metadata.user-agent")
         # java does not fill this tag; it uses the normal http tags
 
-    if not user_agent:
+    if not user_agent and metrics.get("_dd.top_level") == 1.0:
```
Comment (pawelchcki) [inline utils/interfaces/_core.py:181]: It would be good to add comment explaining the new check

**Review by pawelchcki — CHANGES_REQUESTED** [review]
Tiny nit otherwise looks good to me

**Review by pawelchcki — APPROVED** [review]

File: utils/interfaces/_core.py (hunk 1 above)
Comment (lloeki) [reply utils/interfaces/_core.py:181]: This kind of works, but it feels indirect:

- there can be multiple top level tags in the trace (e.g with request queuing synthetic spans, which are in fact at the root of the trace). we don't run in this way but we should have weblog cases for that situation.
- if this is squarely looking for the appsec case then `_dd.appsec.enabled =1.0` is a better, unambiguous marker
- it also begs into question why appsec and tracing duplicate tagging in different places

That said, I'd rather have the CI green again, so OK to merge and fix later while we think about it (cc @marcotc)

**Comment by lloeki** [comment]
Rerunning CI after the recent merge on dd-trace-rb master

**Comment by lloeki** [comment]
Ruby is all green, the other failures are flakiness

**Pull request merged**